service/backend/pchain/indexer/parser.go

```diff
@@ -0,0 +1,329 @@
+	}
+
+	hashID, err := ids.FromString(hash)
+	if err != nil {
+		return nil, err
+	}
+
+	container, err := p.pChainClient.GetContainerByID(ctx, hashID)
+	if err != nil {
+		return nil, err
+	}
+
+	return p.parseBlockBytes(container.Bytes)
+}
+
+func (p *parser) parseBlockBytes(proposerBytes []byte) (*ParsedBlock, error) {
+	errs := wrappers.Errs{}
+
+	proposer, bytes, err := getProposerFromBytes(proposerBytes)
+	if err != nil {
+		return nil, fmt.Errorf("fetching proposer from block bytes errored with %w", err)
+	}
+
+	if p.genesisTimestamp.IsZero() {
+		return nil, errParserUninitialized
+	}
+
+	blk, err := blocks.Parse(p.codec, bytes)
+	if err != nil {
+		return nil, fmt.Errorf("unmarshaling block bytes errored with %w", err)
+	}
+
+	parsedBlock := ParsedBlock{
+		Height:    blk.Height(),
+		BlockID:   blk.ID(),
+		BlockType: fmt.Sprintf("%T", blk),
+		Proposer:  proposer,
+	}
+
+	blockTimestamp := time.Time{}
+
+	switch castBlk := blk.(type) {
+	case *blocks.ApricotProposalBlock:
+		errs.Add(p.initializeTx(castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+
+		// If the block has an advance time tx, use its timestamp as the block timestamp
+		for _, tx := range parsedBlock.Txs {
+			if att, ok := tx.Unsigned.(*txs.AdvanceTimeTx); ok {
+				blockTimestamp = att.Timestamp()
+				break
+			}
+		}
+	case *blocks.BlueberryProposalBlock:
+		errs.Add(p.initializeTx(castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+		blockTimestamp = castBlk.Timestamp()
+	case *blocks.ApricotAtomicBlock:
+		errs.Add(p.initializeTx(castBlk.Tx))
+
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{castBlk.Tx}
+	case *blocks.ApricotStandardBlock:
+		for _, tx := range castBlk.Transactions {
+			errs.Add(p.initializeTx(tx))
+		}
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Transactions
+
+	case *blocks.BlueberryStandardBlock:
+		for _, tx := range castBlk.Transactions {
+			errs.Add(p.initializeTx(tx))
+		}
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = castBlk.Transactions
+		blockTimestamp = castBlk.Timestamp()
+	case *blocks.ApricotAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	case *blocks.BlueberryAbortBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+		blockTimestamp = castBlk.Timestamp()
+	case *blocks.BlueberryCommitBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+		blockTimestamp = castBlk.Timestamp()
+	case *blocks.ApricotCommitBlock:
+		parsedBlock.ParentID = castBlk.PrntID
+		parsedBlock.Txs = []*txs.Tx{}
+	default:
+		errs.Add(fmt.Errorf("no handler exists for block type %T", castBlk))
+	}
+
+	if blockTimestamp.IsZero() {
+		if proposer.Timestamp > genesisTimestamp {
+			blockTimestamp = time.Unix(proposer.Timestamp, 0)
+		} else {
+			blockTimestamp = time.Unix(genesisTimestamp, 0)
+		}
+	}
```

Comment:
The idea here is to return the block timestamp if one was found; if not, then return proposer level timestamp. I wasn't sure if either has to exist, so I'm falling back to the genesis timestamp as the default.

It is technically incorrect since we simply don't have a clue about the timestamp, but if we return 0 instead, then the Rosetta validation complains about it. I'll add a comment to explain the context.